modules/shared/src/main/scala/io/constellationnetwork/currency/http/Decoders.scala

```diff
@@ -0,0 +1,84 @@
+package io.constellationnetwork.currency.http
+
+import cats.data.NonEmptyList
+import cats.effect.Async
+
+import io.constellationnetwork.currency.dataApplication.DataTransaction.DataTransactions
+import io.constellationnetwork.currency.dataApplication._
+import io.constellationnetwork.security.Hashed
+import io.constellationnetwork.security.signature.Signed
+
+import derevo.circe.magnolia.decoder
+import derevo.derive
+import io.circe.syntax.EncoderOps
+import io.circe.{Decoder, Json}
+import org.http4s._
+import org.http4s.circe._
+
+object Decoders {
+
+  def feeTransactionRequestDecoder[F[_]: Async, D <: DataUpdate](
+    implicit dataUpdateDecoder: Decoder[D],
+    feeTransactionDecoder: Decoder[FeeTransaction]
+  ): EntityDecoder[F, DataTransactions] = {
+    @derive(decoder)
+    case class DataTransactionRequest(
+      data: Signed[D],
+      fee: Option[Signed[FeeTransaction]]
+    )
+
+    jsonOf[F, DataTransactionRequest].flatMapR { req =>
+      val signedTransactions: List[Signed[DataTransaction]] = List(req.data) ++ req.fee.toList
+      NonEmptyList.fromList(signedTransactions) match {
+        case Some(nonEmptyTransactions) =>
+          DecodeResult.successT(nonEmptyTransactions)
+        case None =>
+          DecodeResult.failureT(InvalidMessageBodyFailure("Empty list of transactions"))
+      }
+    }
+  }
+
+  def feeTransactionResponseDecoder[F[_]: Async](
+    dataResponse: DataResponse,
+    validationResult: Either[DataApplicationValidationError, NonEmptyList[Hashed[DataTransaction]]]
+  ): F[Response[F]] = {
+    def createErrorResponse(errorMsg: String): F[Response[F]] =
+      Async[F].pure(
+        Response(Status.BadRequest).withEntity(Json.obj("error" -> Json.fromString(errorMsg)))
+      )
+
+    def createResponseData(hashedTransactions: NonEmptyList[Hashed[DataTransaction]]): Json =
+      dataResponse match {
+        case _: SingleDataUpdateResponse =>
+          Map(
+            "hash" -> Json.fromString(hashedTransactions.head.hash.value),
+            "feeHash" -> Json.Null
+          ).asJson
```

Comment:
it's better not to manually create jsons but rely on encoders/decoders. there is already decoder for Map, so you can use it instead of manually creating json.